buildSrc/src/main/kotlin/java-repackage-jars.gradle.kts

```diff
@@ -0,0 +1,42 @@
+import java.util.Calendar
+import java.util.GregorianCalendar
+import java.util.jar.JarEntry
+import java.util.jar.JarFile
+import java.util.jar.JarOutputStream
+import org.gradle.api.internal.file.archive.ZipCopyAction
+
+// This registers a `doLast` action to rewrite the timestamps of the project's output JAR
+afterEvaluate {
+	val jarTask = (tasks.findByName("shadowJar") ?: tasks["jar"]) as Jar
+
+	jarTask.doLast {
+
+		val newFile = createTempFile("rewrite-timestamp")
+		val originalOutput = jarTask.archiveFile.get().getAsFile()
+
+		newFile.outputStream().use { os ->
+
+			val newJarStream = JarOutputStream(os)
+			val oldJar = JarFile(originalOutput)
+
+			oldJar.entries()
+					.toList()
+					.distinctBy { it.name }
+					.sortedBy { it.name }
+					.forEach { entry ->
+						val jarEntry = JarEntry(entry.name)
+
+						// Use the same constant as the fixed timestamps in normal copy actions
+						jarEntry.time = ZipCopyAction.CONSTANT_TIME_FOR_ZIP_ENTRIES
```

Comment:
@mrwilson , this seems to specify the same constant value for all the jars. Is `isPreserveFileTimestamps = false` enough then? What is the purpose of rewriting the jar?

Reply:
This is just for two specific cases where the actual `jar` command was being used to manipulate the artifact after it's been created by Gradle - I tried to replicate what the `jar` tool was doing in pure Gradle, with no success.

As an immediate solution, this resets all the timestamps of the two artifacts in question - the choice of timestamp was informed by what the `isPreserveFileTimestamps = false` would do if this step wasn't necessary.

I hope that helps.

Reply:
Thanks, that clarifies things. It would probably be great if something like that was in the comment above.